src/bank.rs

```diff
@@ -912,9 +913,60 @@ impl Bank {
 
     /// Process an ordered list of entries.
     pub fn process_entries(&self, entries: &[Entry]) -> Result<()> {
+        self.par_process_entries(entries)
+    }
+
+    pub fn all_ok(results: &[Result<()>]) -> Result<()> {
+        for r in results {
```

Comment:
Seems like someone would have created a library function for this, and it'd look something like this:

```rust
pub fn first_err<T: Default, E>(results: Vec<Result<T, E>>) -> Result<T, E> {
     results.into_iter().fold(Ok(Default::default()), Result.and)
}
```

Reply:
```rust
#[derive(Debug, PartialEq)]
enum E {
    E0,
    E1,
}

#[test]
fn test_first_err() {
    assert_eq!(first_err::<(), E>(vec![]), Ok(()));
    assert_eq!(first_err::<(), E>(vec![Ok(())]), Ok(()));
    assert_eq!(first_err::<(), _>(vec![Err(E::E0)]), Err(E::E0));
    assert_eq!(first_err::<(), _>(vec![Err(E::E0), Err(E::E1)]), Err(E::E0));
}
```

Reply:
+1 for the first_err name, I was going to comment as well all_ok doesn't indicate what it does that well.

Reply:
Sunday fun day

```rust
trait ResultIterator<T, E> {
    fn first_err(self) -> Result<T, E>;
}

impl<I, T, E> ResultIterator<T, E> for I
where
    I: IntoIterator<Item = Result<T, E>>,
    T: Default,
{
    fn first_err(self) -> Result<T, E> {
        self.into_iter().fold(Ok(Default::default()), Result::and)
    }
}

#[derive(Debug, PartialEq)]
enum E {
    E0,
    E1,
}

#[test]
fn test_first_err() {
    let empty: Vec<Result<_, E>> = vec![];
    assert_eq!(empty.first_err(), Ok(()));

    let one_ok: Vec<Result<_, E>> = vec![Ok(())];
    assert_eq!(one_ok.first_err(), Ok(()));

    let one_err: Vec<Result<(), _>> = vec![Err(E::E0)];
    assert_eq!(one_err.first_err(), Err(E::E0));

    let two_errs: Vec<Result<(), _>> = vec![Err(E::E0), Err(E::E1)];
    assert_eq!(two_errs.first_err(), Err(E::E0));

    assert_eq!(vec![Ok(()), Err(E::E0)].first_err(), Err(E::E0));
}
```

Reply:
So we never need to think about this again, can you add this to result.rs:

```rust
pub fn first_err<I, T, E>(xs: I, def: T) -> Result<T, E>
where
    I: IntoIterator<Item = Result<T, E>>,
{
    for x in xs {
        x?;
    }
    Ok(def)
}

#[derive(Debug, PartialEq)]
enum E {
    E0,
    E1,
}

#[test]
fn test_first_err() {
    assert_eq!(first_err::<_, _, E>(vec![], ()), Ok(()));
    assert_eq!(first_err::<_, _, E>(vec![Ok(())], ()), Ok(()));
    assert_eq!(first_err(vec![Err(E::E0)], ()), Err(E::E0));
    assert_eq!(first_err(vec![Err(E::E0), Err(E::E1)], ()), Err(E::E0));
    assert_eq!(first_err(vec![Ok(()), Err(E::E0)], ()), Err(E::E0));
}
```

Reply:
@garious want me to use this?  or you think its overkill?

Reply:
Use it.  It's not overkill. Add it to result.rs.  I've used this `for x in xs { x? }` pattern a dozen times, and I'd like to finally have a function for it.